components/raftstore/src/store/peer.rs

```diff
@@ -1064,6 +1064,23 @@ impl Peer {
         false
     }
 
+    /// Whether a log can be applied before writing raft batch.
+    ///
+    /// If TiKV crashes, it's possible apply index > commit index. If logs are still
+    /// available in other nodes, it's possible to be recovered. But for singleton, logs are
+    /// only available on single node, logs are gone forever.
+    ///
+    /// Note we can't just check singleton. Because conf change takes effect on apply, so even
+    /// there are two nodes, previous logs can still be committed by leader alone. Those logs
+    /// can't be applied early. After introducing joint consensus, the node number can be
+    /// undetermined. So here check whether log is persisted on disk instead.
+    ///
+    /// Only apply existing logs has another benefit that we don't need to deal with snapshots
+    /// that are older than apply index as apply index <= last index <= index of snapshot.
+    pub fn can_early_apply(&self, term: u64, index: u64) -> bool {
+        self.get_store().last_index() >= index && self.get_store().last_term() >= term
```

Comment:
I think we can use the first index of `entries` in raft ready to judge if the committed entries are stable. 
Actually, this condition can be more relaxed, we can just send the stable entries to apply fsm and set the commit index to the index of the last stable entry.
But I think this optimization is only useful in the leader, so maybe it's ok to not relax condition.

Reply:
> ...use the first index of entries in raft ready to judge...

How?

> ...just send the stable entries to apply fsm and set the commit index to the index of the last stable entry...

I don't know what does "stable" mean. If it just means entries written to disk then it's wrong to assume last stable entries are committed. For example, an entry can be committed after several round of handle ready.